Parameter to add exclude patterns

Some users might want linuxdeploy to skip certain libraries in addition to the official [excludelist](https://github.com/AppImage/AppImages/blob/master/excludelist). We should add a parameter for that purpose.

I'm trying to include JDK in my AppImage, and it fails to deploy with this:
```
-- Deploying dependencies for existing files in AppDir -- 
Deploying dependencies for ELF file install/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/jexec 
Skipping deployment of blacklisted library /lib/x86_64-linux-gnu/libc.so.6 
Deploying dependencies for ELF file install/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/amd64/libsaproc.so 
Skipping deployment of blacklisted library /lib/x86_64-linux-gnu/libthread_db.so.1 
Skipping deployment of blacklisted library /lib/x86_64-linux-gnu/libdl.so.2 
Deploying dependencies for ELF file install/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/amd64/libunpack.so 
ERROR: Could not find dependency: libjvm.so 
ERROR: Failed to deploy dependencies for existing files 
Error: Process completed with exit code 1.
```
The JDK is installed via apt-get and the folder is copied over.

--- a/include/linuxdeploy/core/appdir.h
+++ b/include/linuxdeploy/core/appdir.h
@@ -38,6 +38,9 @@ namespace linuxdeploy {
                     // shortcut for using a normal string instead of a path
                     explicit AppDir(const std::string& path);
 
+                    // Set additional shared library name patterns to be excluded from deployment.
+                    void setExcludeLibraryPatterns(const std::vector<std::string> &excludeLibraryPatterns);
+
                     // creates basic directory structure of an AppDir in "FHS" mode
                     bool createBasicStructure() const;
 

--- a/src/core/appdir.cpp
+++ b/src/core/appdir.cpp
@@ -98,6 +98,7 @@ namespace linuxdeploy {
             class AppDir::PrivateData {
                 public:
                     bf::path appDirPath;
+                    std::vector<std::string> excludeLibraryPatterns;
 
                     // store deferred operations
                     // these can be executed by calling excuteDeferredOperations
@@ -121,7 +122,7 @@ namespace linuxdeploy {
                     bool disableCopyrightFilesDeployment = false;
 
                 public:
-                PrivateData() : copyOperationsStorage(), stripOperations(), setElfRPathOperations(), visitedFiles(), appDirPath() {
+                PrivateData() : copyOperationsStorage(), stripOperations(), setElfRPathOperations(), visitedFiles(), appDirPath(), excludeLibraryPatterns() {
                         copyrightFilesManager = copyright::ICopyrightFilesManager::getInstance();
                     };
 
@@ -392,8 +393,8 @@ namespace linuxdeploy {
                             return false;
                         }
 
-                        static auto isInExcludelist = [](const bf::path& fileName) {
-                            for (const auto& excludePattern : generatedExcludelist) {
+                        static auto isInExcludelist = [](const bf::path& fileName, const std::vector<std::string> &excludeList) {
+                            for (const auto& excludePattern : excludeList) {
                                 // simple string match is faster than using fnmatch
                                 if (excludePattern == fileName)
                                     return true;
@@ -413,7 +414,7 @@ namespace linuxdeploy {
                             return false;
                         };
 
-                        if (!forceDeploy && isInExcludelist(path.filename())) {
+                        if (!forceDeploy && (isInExcludelist(path.filename(), generatedExcludelist) || isInExcludelist(path.filename(), excludeLibraryPatterns))) {
                             ldLog() << "Skipping deployment of blacklisted library" << path << std::endl;
 
                             // mark file as visited
@@ -638,6 +639,10 @@ namespace linuxdeploy {
             }
 
             AppDir::AppDir(const std::string& path) : AppDir(bf::path(path)) {}
+
+            void AppDir::setExcludeLibraryPatterns(const std::vector<std::string> &excludeLibraryPatterns) {
+                d->excludeLibraryPatterns = excludeLibraryPatterns;
+            }
 
             bool AppDir::createBasicStructure() const {
                 std::vector<std::string> dirPaths = {

--- a/src/main.cpp
+++ b/src/main.cpp
@@ -33,6 +33,7 @@ int main(int argc, char** argv) {
     args::ValueFlag<std::string> appDirPath(parser, "appdir", "Path to target AppDir", {"appdir"});
 
     args::ValueFlagList<std::string> sharedLibraryPaths(parser, "library", "Shared library to deploy", {'l', "library"});
+    args::ValueFlagList<std::string> excludeLibraryPatterns(parser, "pattern", "Shared library to exclude from deployment (glob pattern)", {"exclude-library"});
 
     args::ValueFlagList<std::string> executablePaths(parser, "executable", "Executable to deploy", {'e', "executable"});
 
@@ -110,6 +111,7 @@ int main(int argc, char** argv) {
     }
 
     appdir::AppDir appDir(appDirPath.Get());
+    appDir.setExcludeLibraryPatterns(excludeLibraryPatterns.Get());
 
     // allow disabling copyright files deployment via environment variable
     if (getenv("DISABLE_COPYRIGHT_FILES_DEPLOYMENT") != nullptr) {

